File: frontend/src/lib/canvasOps.ts
/**
 * Canvas Operations Protocol — client-side handler.
 * Receives SSE events from backend, applies them to Excalidraw.
 */

export type OpType =
  | "create_note"
  | "create_text"
  | "create_diagram"
  | "create_sticky"
  | "update_element"
  | "move_element"
  | "delete_element"
  | "group_elements"
  | "create_edge_line"
  | "set_background"
  | "set_theme"
  | "pan_to"
  | "zoom_to"
  | "stream_start"
  | "stream_chunk"
  | "stream_end"
  | "arrange_cluster"
  | "batch"
  | "info"
  | "error"
  | "done"

export interface CanvasOp {
  op: OpType
  element_id?: string
  x?: number
  y?: number
  width?: number
  height?: number
  text?: string
  color?: string
  theme?: string
  zoom?: number
  style?: string
  note?: Record<string, unknown>
  note_id?: string
  elements?: Record<string, unknown>[]
  connections?: Record<string, unknown>[]
  operations?: CanvasOp[]
  topology?: Record<string, unknown>
  message?: string
  metadata?: Record<string, unknown>
  timestamp?: number
}

export interface SSEEvent {
  event: string
  data: Record<string, unknown>
}

export interface Viewport {
  x: number
  y: number
  width: number
  height: number
  zoom: number
}

export interface StreamRequest {
  message: string
  viewport?: Viewport
  history?: Array<{ role: string; content: string }>
  selected_element_ids?: string[]
  context_type?: string
}

export interface StreamCallbacks {
  onIntent?: (
    intent: string,
    topic: string,
    metadata: Record<string, unknown>
  ) => void
  onChat?: (content: string) => void
  onCanvasOp?: (op: CanvasOp) => void
  onSources?: (
    sources: Array<{ id: string; title: string; similarity: number }>
  ) => void
  onFollowUps?: (followUps: string[]) => void
  onError?: (message: string) => void
  onDone?: () => void
}

const API_BASE = import.meta.env.VITE_API_URL || "http://localhost:8000/api"

function normalizeApiBase(base: string): string {
  return base.endsWith("/") ? base.slice(0, -1) : base
}

function getAuthHeader(): Record<string, string> | null {
  const token =
    localStorage.getItem("mnemos-token") ||
    localStorage.getItem("access_token")
  if (token && token !== "auth-disabled") {
    return { Authorization: `Bearer ${token}` }
  }
  return null
}

/**
 * Stream canvas operations from the backend via SSE.
 * Returns an AbortController so the caller can cancel.
 */
export function streamCanvasOps(
  pageId: string,
  request: StreamRequest,
  callbacks: StreamCallbacks,
  apiBase: string = API_BASE
): AbortController {
  const controller = new AbortController()
  const baseUrl = normalizeApiBase(apiBase)

  const run = async () => {
    try {
      const response = await fetch(`${baseUrl}/canvas/${pageId}/stream`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          ...(getAuthHeader() || {}),
        },
        body: JSON.stringify(request),
        signal: controller.signal,
      })

      if (!response.ok) {
        callbacks.onError?.(`HTTP ${response.status}: ${response.statusText}`)
        callbacks.onDone?.()
        return
      }

      const reader = response.body?.getReader()
      if (!reader) {
        callbacks.onError?.("No response body")
        callbacks.onDone?.()
        return
      }

      const decoder = new TextDecoder()
      let buffer = ""
      let currentEvent = ""
      let currentDataLines: string[] = []

      const flushEvent = () => {
        if (!currentEvent) return

        const payload = currentDataLines.join("\n")
        if (!payload) {
          currentEvent = ""
          currentDataLines = []
          return
        }

        try {
          const parsed = JSON.parse(payload)
          dispatchEvent(currentEvent, parsed, callbacks)
        } catch {
          console.warn("Failed to parse SSE data:", payload)
        }

        currentEvent = ""
        currentDataLines = []
      }

      while (true) {
        const { done, value } = await reader.read()
        if (done) break

        buffer += decoder.decode(value, { stream: true })
        const lines = buffer.split("\n")
        buffer = lines.pop() || ""

        for (const line of lines) {
          const normalized = line.endsWith("\r") ? line.slice(0, -1) : line

          if (!normalized) {
            flushEvent()
            continue
          }

          if (normalized.startsWith("event: ")) {
            currentEvent = normalized.slice(7).trim()
          } else if (normalized.startsWith("data: ")) {
            currentDataLines.push(normalized.slice(6))
          }
        }
      }

      // Flush terminal event
      if (currentEvent && currentDataLines.length > 0) {
        flushEvent()
      }
    } catch (e: unknown) {
      if (e instanceof Error && e.name !== "AbortError") {
        callbacks.onError?.(e.message || "Stream failed")
      }
    }

    callbacks.onDone?.()
  }

  run()
  return controller
}

function dispatchEvent(
  event: string,
  data: Record<string, unknown>,
  callbacks: StreamCallbacks
) {
  switch (event) {
    case "intent":
      callbacks.onIntent?.(
        data.intent as string,
        data.topic as string,
        (data.metadata as Record<string, unknown>) || {}
      )
      break
    case "chat":
      callbacks.onChat?.(data.content as string)
      break
    case "canvas_op":
      callbacks.onCanvasOp?.(data as unknown as CanvasOp)
      break
    case "sources":
      callbacks.onSources?.(
        (data.sources as Array<{
          id: string
          title: string
          similarity: number
        }>) || []
      )
      break
    case "follow_ups":
      callbacks.onFollowUps?.((data.follow_ups as string[]) || [])
      break
    case "error":
      callbacks.onError?.((data.message as string) || "Unknown error")
      break
    case "done":
      break
  }
}

File: backend/app-old/models/canvas_ops.py
# === FILE: backend/app/models/canvas_ops.py ===
"""
Canvas Operations Protocol.

Every backend action that touches the canvas is expressed as a CanvasOp.
The frontend receives these over SSE and applies them to Excalidraw.
"""

from __future__ import annotations
from enum import Enum
from typing import Any, Optional
from pydantic import BaseModel, Field
import uuid
import time


class OpType(str, Enum):
    # Element CRUD
    CREATE_NOTE = "create_note"
    CREATE_TEXT = "create_text"
    CREATE_DIAGRAM = "create_diagram"
    CREATE_STICKY = "create_sticky"
    UPDATE_ELEMENT = "update_element"
    MOVE_ELEMENT = "move_element"
    DELETE_ELEMENT = "delete_element"
    GROUP_ELEMENTS = "group_elements"
    CREATE_EDGE_LINE = "create_edge_line"

    # Canvas state
    SET_BACKGROUND = "set_background"
    SET_THEME = "set_theme"
    PAN_TO = "pan_to"
    ZOOM_TO = "zoom_to"

    # Streaming text
    STREAM_START = "stream_start"
    STREAM_CHUNK = "stream_chunk"
    STREAM_END = "stream_end"

    # Batch / arrange
    ARRANGE_CLUSTER = "arrange_cluster"
    BATCH = "batch"

    # Meta
    INFO = "info"
    ERROR = "error"
    DONE = "done"


class CanvasOp(BaseModel):
    op: OpType
    element_id: Optional[str] = None
    x: Optional[float] = None
    y: Optional[float] = None
    width: Optional[float] = None
    height: Optional[float] = None
    text: Optional[str] = None
    color: Optional[str] = None
    theme: Optional[str] = None
    zoom: Optional[float] = None
    style: Optional[str] = None
    note: Optional[dict] = None
    note_id: Optional[str] = None
    elements: Optional[list[dict]] = None
    connections: Optional[list[dict]] = None
    operations: Optional[list["CanvasOp"]] = None
    topology: Optional[dict] = None
    message: Optional[str] = None
    metadata: dict = Field(default_factory=dict)
    timestamp: int = Field(default_factory=lambda: int(time.time() * 1000))


class SSEEvent(BaseModel):
    event: str  # "intent", "chat", "canvas_op", "sources", "follow_ups", "error", "done"
    data: dict


class Viewport(BaseModel):
    x: float = 0
    y: float = 0
    width: float = 1920
    height: float = 1080
    zoom: float = 1.0

    @property
    def center_x(self) -> float:
        return self.x + self.width / 2

    @property
    def center_y(self) -> float:
        return self.y + self.height / 2

    @property
    def right(self) -> float:
        return self.x + self.width

    @property
    def bottom(self) -> float:
        return self.y + self.height

    def contains(self, px: float, py: float) -> bool:
        return self.x <= px <= self.right and self.y <= py <= self.bottom


class Rect(BaseModel):
    x: float
    y: float
    w: float
    h: float

    @property
    def right(self) -> float:
        return self.x + self.w

    @property
    def bottom(self) -> float:
        return self.y + self.h

    @property
    def center_x(self) -> float:
        return self.x + self.w / 2

    @property
    def center_y(self) -> float:
        return self.y + self.h / 2

    def overlaps(self, other: "Rect", gap: float = 0) -> bool:
        return not (
            self.right + gap <= other.x
            or other.right + gap <= self.x
            or self.bottom + gap <= other.y
            or other.bottom + gap <= self.y
        )


class Placement(BaseModel):
    x: float
    y: float
    cluster_id: Optional[str] = None
    strategy: str = "auto"
    reason: str = ""


class CanvasStreamRequest(BaseModel):
    message: str
    viewport: Optional[Viewport] = None
    history: list[dict] = []
    selected_element_ids: list[str] = []
    context_type: str = "page"


class Intent(str, Enum):
    COMPOSE = "compose"
    COMMAND = "command"
    ARRANGE = "arrange"
    CAPTURE = "capture"
    QUERY = "query"
    DIAGRAM = "diagram"
    SEARCH = "search"
    NAVIGATE = "navigate"


def make_element_id(prefix: str = "el") -> str:
    return f"{prefix}_{uuid.uuid4().hex[:12]}"